### PR TITLE
Add output flag to `pack inspect-image`

### LINKED ISSUE #490: pack inspect-image should accept an --output flag
### Description
`pack inspect-image --bom` produces a wall of text that is hard to read. We should make it easier to understand the output of this command.

### Proposed solution
Output the data in a format that I want.

### Possible acceptance criteria

**When** I run `pack inspect-image --output json` or `pack inspect-image -o json`
**Then** I see the output in json format (with BOM information)

**When** I run `pack inspect-image --output yaml` or `pack inspect-image -o yaml`
**Then** I see the output in yaml format (with BOM information)

**When** I run `pack inspect-image --output toml` or `pack inspect-image -o toml`
**Then** I see the output in toml format (with BOM information)

Don't break the existing behavior:

**When** I run `pack inspect-image` (without `--bom`)
**Then** I should not see the BOM displayed

**When** I run `pack inspect-image --bom`
**Then** I see the BOM output in json format

### Describe alternatives you've considered
Do nothing. It's fine as it is.

Output the data in a table (this would be very difficult because the data schema is undefined).

### Additional context
* This is to address the pain point described in this issue: https://github.com/buildpacks/pack/issues/378
* The implementation should follow the pattern in this issue: https://github.com/buildpacks/pack/issues/387
* We should announce that `pack inspect-image --bom` is deprecating and will be going away. 
* Create a separate issue for removing `--bom` later.

### PR DESCRIPTION
**Pull request by dwillist**
## Summary

Allow `pack inspect-image` to accept an `--output` flag & produce outputs in  `yaml` `json` and `toml`.

Additionally adds `yaml` ~`toml`~ formatting for `pack inspect-image <image-name> --bom`,

( the golang `toml` parsing implementation has difficulties marshalling arbitrary structs into toml and so it is omitted)

## Documentation

- Should this change be documented?

    - [ x ] No (I think?)

## Related

Resolves #490

### CODE REVIEW

File: cmd/cmd.go [hunk 1]
```diff
@@ -65,12 +67,14 @@ func NewPackCommand(logger ConfigurableLogger) (*cobra.Command, error) {
 
 	rootCmd.AddCommand(commands.Build(logger, cfg, &packClient))
 	rootCmd.AddCommand(commands.Rebase(logger, cfg, &packClient))
-	rootCmd.AddCommand(commands.InspectImage(logger, &cfg, &packClient))
+	// TODO: is the config the base of truth for what run image mirrors are displayed??
+	//       or do we use the run image mirrors located on the image??
+	rootCmd.AddCommand(commands.InspectImage(logger, writer.NewFactory(), cfg, &packClient))
```
Comment (dfreilich) [inline cmd/cmd.go:72]: When inspecting an image, I would expect the run image mirrors on the image to be the source of truth, while the config is the base for future images built with `pack`.

File: internal/inspectimage/writer/structured_bom_format_test.go [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+						RunImageMirrors: []config.RunImage{},
+					}
+
+					structuredBOMWriter := writer.StructuredBOMFormat{
+						MarshalFunc: testMarshalFunc,
+					}
+
+					logger := ilogging.NewLogWithWriters(&outBuf, &outBuf)
+					err := structuredBOMWriter.Print(logger, sharedImageInfo, nil, nil, nil, nil)
+					assert.ErrorWithMessage(err, fmt.Sprintf("unable to find image '%s' locally or remotely", "missing-image"))
```
Comment (dfreilich) [inline internal/inspectimage/writer/structured_bom_format_test.go:60]: This seems like a slightly different assertion than we were doing. Is this output still being generated? 
https://github.com/buildpacks/pack/pull/957/files#diff-2dbbb729dc74480a1e93e56403bb3246e3e81ba9ba7ae8b0e4486ff321ec45eeL71-L79
```
h.AssertEq(t,
						outBuf.String(),
						`{"remote":null,"local":null}`+"\n")
```

File: internal/inspectimage/info.go [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+	if info == nil {
+		return nil
+	}
+	return displayBOM(info.BOM)
+}
+
+//
+// private functions
+//
+
```
Comment (dfreilich) [inline internal/inspectimage/info.go:91]: nit: Can we move this out to a separate `bom_display.go` file, to make the files a bit more cohesive?

File: internal/inspectimage/info.go [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+func GetConfigMirrors(info *pack.ImageInfo, imageMirrors []config.RunImage) []string {
```
Comment (dfreilich) [inline internal/inspectimage/info.go:92]: nit: Any reason this is public? It looks like it's only used in this file

File: internal/inspectimage/info.go [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+		return nil
+	}
+	return &InfoDisplay{
+		StackID:         info.StackID,
+		Base:            displayBase(info.Base),
+		RunImageMirrors: displayMirrors(info, generalInfo),
+		Buildpacks:      displayBuildpacks(info.Buildpacks),
+		Processes:       displayProcesses(info.Processes),
+	}
+}
```
Comment (dfreilich) [inline internal/inspectimage/info.go:65]: nit: Seems like this is the main concept in this file. Can we rename it to `info_display.go`?

**Review by dfreilich — APPROVED** [review]
This is super super impressive. A buncha questions and small questions, but overall this is 

legendary

File: acceptance/acceptance_test.go [hunk 6]
```diff
@@ -876,38 +876,66 @@ func testAcceptance(
 
 						if pack.Supports("inspect-image") {
 							t.Log("inspect-image")
-							output = pack.RunSuccessfully("inspect-image", repoName)
 
 							var (
 								webCommand   string
 								helloCommand string
-								helloArgs    string
+								helloArgs    []string
```
[the remark this answers is not in the document]
Comment (dwillist) [reply acceptance/acceptance_test.go:883]: Ya exactly.

File: cmd/cmd.go (hunk 1 above)
Comment (dwillist) [reply cmd/cmd.go:72]: Ya comment should be removed, as things stand we print run images from Both, 
- run images specified in the `config` are indicated by a `(user-provided)` field.

File: internal/inspectimage/writer/structured_bom_format_test.go (hunk 2 above)
Comment (dwillist) [reply internal/inspectimage/writer/structured_bom_format_test.go:60]: Ya this is different behavior. Could revert if we want to keep a `pack` consistent. But it does feel awkward having `pack inspect-builder <image-name> --bom` succeed and return an empty `json` array when we failed to find relevant images.

Comment (dwillist) [reply internal/inspectimage/writer/structured_bom_format_test.go:60]: Ya this is different behavior. Could revert if we want to keep a `pack` consistent. But it does feel awkward having `pack inspect-builder <image-name> --bom` succeed and return an empty `json` array when we failed to find relevant images.

**Comment by jromero** [comment]
# ✅ Accepted

```
❯ pack-dev inspect-image --help
Usage:
  pack inspect-image <image-name> [flags]

Examples:
pack inspect-image buildpacksio/pack

Flags:
      --bom             print bill of materials
  -h, --help            Help for 'inspect-image'
  -o, --output string   Output format to display builder detail (json, yaml, toml, human-readable).
                        Omission of this flag will display as human-readable. (default "human-readable")

Global Flags:
      --no-color     Disable color output
  -q, --quiet        Show less output
      --timestamps   Enable timestamps in output
  -v, --verbose      Show more output


❯ pack-dev inspect-image my-app
Inspecting image: my-app

REMOTE:
(not present)

LOCAL:

Stack: io.buildpacks.samples.stacks.bionic

Base Image:
  Reference: cfd856639a34e5ef0aab2772a5a7310fa5fb9025a664705173263910425ff0a0
  Top Layer: sha256:753c7b8b885ec1120a5380a02e5299f69488cb193aa5a89f753f38926f86c86e

Run Images:
  index.docker.io/cnbs/sample-stack-run:bionic         (user-configured)
  index.docker.io/cnbs/sample-stack-run:bionic1        (user-configured)
  cnbs/sample-stack-run:bionic

Buildpacks:
  ID                         VERSION
  samples/bash-script        0.0.1

Processes:
  TYPE                 SHELL        COMMAND        ARGS
  web (default)        bash         ./app.sh


❯ pack-dev inspect-image my-app -o json
{
  "image_name": "my-app",
  "remote_info": null,
  "local_info": {
    "stack": "io.buildpacks.samples.stacks.bionic",
    "base_image": {
      "top_layer": "sha256:753c7b8b885ec1120a5380a02e5299f69488cb193aa5a89f753f38926f86c86e",
      "reference": "cfd856639a34e5ef0aab2772a5a7310fa5fb9025a664705173263910425ff0a0"
    },
    "run_images": [
      {
        "name": "index.docker.io/cnbs/sample-stack-run:bionic",
        "user_configured": true
      },
      {
        "name": "index.docker.io/cnbs/sample-stack-run:bionic1",
        "user_configured": true
      },
      {
        "name": "cnbs/sample-stack-run:bionic"
      }
    ],
    "buildpacks": [
      {
        "id": "samples/bash-script",
        "version": "0.0.1"
      }
    ],
    "processes": [
      {
        "type": "web",
        "shell": "bash",
        "command": "./app.sh",
        "default": true,
        "args": null
      }
    ]
  }
}


❯ pack-dev inspect-image my-app -o toml
image_name = "my-app"

[local_info]
  stack = "io.buildpacks.samples.stacks.bionic"

  [local_info.base_image]
    reference = "cfd856639a34e5ef0aab2772a5a7310fa5fb9025a664705173263910425ff0a0"
    top_layer = "sha256:753c7b8b885ec1120a5380a02e5299f69488cb193aa5a89f753f38926f86c86e"

  [[local_info.buildpacks]]
    id = "samples/bash-script"
    version = "0.0.1"

  [[local_info.processes]]
    args = []
    command = "./app.sh"
    default = true
    shell = "bash"
    type = "web"

  [[local_info.run_images]]
    name = "index.docker.io/cnbs/sample-stack-run:bionic"
    user_configured = true

  [[local_info.run_images]]
    name = "index.docker.io/cnbs/sample-stack-run:bionic1"
    user_configured = true

  [[local_info.run_images]]
    name = "cnbs/sample-stack-run:bionic"


❯ pack-dev inspect-image my-app -o yaml
image_name: my-app
remote_info: null
local_info:
    stack: io.buildpacks.samples.stacks.bionic
    base_image:
        top_layer: sha256:753c7b8b885ec1120a5380a02e5299f69488cb193aa5a89f753f38926f86c86e
        reference: cfd856639a34e5ef0aab2772a5a7310fa5fb9025a664705173263910425ff0a0
    run_images:
        - name: index.docker.io/cnbs/sample-stack-run:bionic
          user_configured: true
        - name: index.docker.io/cnbs/sample-stack-run:bionic1
          user_configured: true
        - name: cnbs/sample-stack-run:bionic
    buildpacks:
        - id: samples/bash-script
          version: 0.0.1
    processes:
        - type: web
          shell: bash
          command: ./app.sh
          default: true
          args: []
```

**Review by jromero — APPROVED** [review]

**Pull request merged**